xarray/tests/test_concat.py

```diff
@@ -46,14 +129,332 @@ def test_concat_compat() -> None:
 
     for var in ["has_x", "no_x_y"]:
         assert "y" not in result[var].dims and "y" not in result[var].coords
-    with pytest.raises(
-        ValueError, match=r"coordinates in some datasets but not others"
-    ):
+    with pytest.raises(ValueError, match=r"'q' not present in all datasets"):
         concat([ds1, ds2], dim="q")
-    with pytest.raises(ValueError, match=r"'q' is not present in all datasets"):
+    with pytest.raises(ValueError, match=r"'q' not present in all datasets"):
         concat([ds2, ds1], dim="q")
 
 
+def test_concat_missing_var() -> None:
+    datasets = create_concat_datasets(2, 123)
```

Comment:
This used to work, because `foo` was not in the first dataset. It would have raised an error if `foo` was in the first dataset. 

Now `foo` gets reindexed too, so I've added a third test below, which covers for that.